discovery/eureka/eureka.go

```diff
@@ -0,0 +1,194 @@
+type Discovery struct {
+	*refresh.Discovery
+	client     *http.Client
+	server     string
+	appsClient applicationsClient
+}
+
+// New creates a new Eureka discovery for the given role.
+func NewDiscovery(conf SDConfig, logger log.Logger) (*Discovery, error) {
+	rt, err := config_util.NewRoundTripperFromConfig(conf.HTTPClientConfig, "eureka_sd", false)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Discovery{
+		client:     &http.Client{Transport: rt},
+		server:     conf.Server,
+		appsClient: fetchApps,
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"eureka",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	targetMap, err := d.fetchTargetGroups(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	all := make([]*targetgroup.Group, 0, len(targetMap))
+	for _, tg := range targetMap {
+		all = append(all, tg)
+	}
+
+	return all, nil
+}
+
+func (d *Discovery) fetchTargetGroups(ctx context.Context) (map[string]*targetgroup.Group, error) {
+	apps, err := d.appsClient(ctx, d.server, d.client)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := appsToTargetGroups(apps)
+	return groups, nil
+}
+
+// appsToTargetGroups takes an array of Eureka Applications and converts them into target groups.
+func appsToTargetGroups(apps *Applications) map[string]*targetgroup.Group {
+	tgroups := map[string]*targetgroup.Group{}
+	for _, a := range apps.Applications {
+		group := createTargetGroup(&a)
+		tgroups[group.Source] = group
+	}
+	return tgroups
+}
+
+func createTargetGroup(app *Application) *targetgroup.Group {
+	var (
+		targets = targetsForApp(app)
+		appName = model.LabelValue(app.Name)
+	)
+	tg := &targetgroup.Group{
+		Targets: targets,
+		Labels: model.LabelSet{
+			appNameLabel: appName,
+		},
+		Source: "eureka",
+	}
+
+	return tg
+}
+
+func targetsForApp(app *Application) []model.LabelSet {
+	targets := make([]model.LabelSet, 0, len(app.Instances))
+
+	// Gather info about the app's 'instances'. Each instance is considered a task
+	for _, t := range app.Instances {
+		targetAddress := targetEndpoint(&t)
+		target := model.LabelSet{
+			model.AddressLabel:  model.LabelValue(targetAddress),
+			model.InstanceLabel: model.LabelValue(t.InstanceID),
+
+			appInstanceHostNameLabel: lv(t.HostName),
+			appInstanceIPAddrLabel:   lv(t.IPAddr),
+			appInstanceStatusLabel:   lv(t.Status),
+			appInstanceIDLabel:       lv(t.InstanceID),
+		}
+
+		if t.Metadata != nil {
+			for k, v := range t.Metadata.Map {
+				ln := strutil.SanitizeLabelName(k)
+				target[model.LabelName(appInstanceMetadataPrefix+ln)] = lv(v)
+			}
+		}
+
+		targets = append(targets, target)
+	}
+	return targets
+}
+
+func lv(s string) model.LabelValue {
+	return model.LabelValue(s)
+}
+
+func targetEndpoint(instance *Instance) string {
+	var port string
+	if instance.Metadata != nil &&
+		len(instance.Metadata.Map["management.port"]) > 0 {
```

Comment:
This seems like business logic to me. It should not go in this SD.

Reply:
It is a feature of a spring-boot. Eureka is often used as a spring-boot base.
The spring-boot uses an `actuator` to create metric endpoint. By default, metric endpoint are created using the port on which they are serviced. Developer can change the port on metric endpoint. That's the management port. When a management port is set up, the metadata of Eureka automatically contains that information. That is `management.port`.
I think it's convenient to have that logic, what do you think?

Reply:
To me it looks like it would be better to have that in the exemple documentation then, as a relabel config.

WDYT @brian-brazil @cstyan

Reply:
Indeed, this sounds like business logic.

Reply:
This is still here

Reply:
I'll delete this logic.